packages/web-extension/src/walletManager/walletManager.ts

```diff
@@ -0,0 +1,388 @@
+export class WalletManager<Metadata extends { name: string }> implements WalletManagerApi {
+  activeWalletId$: Observable<WalletManagerActivateProps>;
+  activeWallet$ = new BehaviorSubject<ObservableWallet | null>(null);
+
+  #activeWalletProps: WalletManagerActivateProps | null = null;
+  #walletStores = new Map<string, storage.WalletStores>();
+  #walletFactory: WalletFactory;
+  #storesFactory: StoresFactory;
+  #walletRepository: WalletRepository<Metadata>;
+  #signerManagerApi: SignerManagerSignApi<Metadata>;
+  #logger: Logger;
+  #managerStorageKey: string;
+  #managerStorage: Storage.StorageArea;
+  #activeWalletId$ = new ReplaySubject<WalletManagerActivateProps>(1);
+
+  constructor(
+    { name }: WalletManagerProps,
+    {
+      walletFactory,
+      storesFactory,
+      logger,
+      managerStorage,
+      walletRepository,
+      signerManagerApi
+    }: MessengerDependencies & WalletManagerDependencies<Metadata>
+  ) {
+    this.#walletRepository = walletRepository;
+
+    this.activeWalletId$ = this.#activeWalletId$.pipe(distinctUntilChanged(deepEquals));
+
+    this.#walletFactory = walletFactory;
+    this.#managerStorageKey = `${name}-activate`;
+    this.#managerStorage = managerStorage;
+    this.#storesFactory = storesFactory;
+    this.#logger = logger;
+    this.#signerManagerApi = signerManagerApi;
+  }
+
+  /**
+   * Switches the network of the active wallet.
+   *
+   * @param id The network id to switch to.
+   */
+  switchNetwork(id: Cardano.ChainId): Promise<void> {
+    if (!this.#hasActiveWallet()) return Promise.resolve();
+
+    const props = { ...this.#activeWalletProps!, chainId: id };
+    return this.activate(props);
+  }
+
+  /** `activate` the wallet with props of last activated wallet (load from `managerStorage`) */
+  async initialize() {
+    let { [this.#managerStorageKey]: lastActivateProps } = await this.#managerStorage.get(this.#managerStorageKey);
+
+    if (!lastActivateProps) return;
+
+    // We need to make sure we are still able to parse the old format of wallet id. Previously we were
+    // storing the following properties:
+    // {
+    //   observableWalletName: 'old-wallet',
+    //   walletId: 'string' <-- I.E '0-888-da7b4795b11a79116eb5232c83d2c862', concatenates chain id and public key hash.
+    //   provider?: { type: P; options: O };
+    // }
+    //
+    // Now we are storing the following properties:
+    // {
+    //   walletId: WalletId; <-- I.E 'da7b4795b11a79116eb5232c83d2c862', only the 16 byte length pub key hash.
+    //   accountIndex?: number; <-- optional account number
+    //   chainId: Cardano.ChainId; <-- I.E { networkId: Cardano.NetworkId.Testnet, networkMagic: 1097911063 }
+    //   provider?: { type: P; options: O };
+    // }
+    //
+    // If we find a legacy wallet id, we need to map it to the new format and save it to storage.
+    if (this.#isLegacyProps(lastActivateProps.walletId)) {
+      lastActivateProps = this.#mapLegacyProps(lastActivateProps);
```

Comment:
This would make it easier to migrate, because Lace stores wallet name in UI context: it will have to create the wallet in repository from the UI.
```suggestion
      lastActivateProps = this.#mapLegacyProps(lastActivateProps);
      if (!activeWalletExistsInRepository(lastActivateProps)) return;
```

Reply:
done